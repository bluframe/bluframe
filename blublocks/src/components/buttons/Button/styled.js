// @flow
/* eslint-disable no-ternary, no-nested-ternary */

import styled, { type StyledComponent } from "styled-components"
import Reset from "components/buttons/Reset"
import { rgba } from "polished"

type Props = {|
  +isBold?: boolean,
  +isPadded?: boolean
|}

const BACKGROUND_ALPHA = 0.7

export const Wrapper: StyledComponent<
  Props,
  mixed,
  React$ComponentType<Props>
> = styled(Reset)`
  background: ${({ theme }) => theme.palette.primary.main};
  border-radius: 3px;
  color: ${({ theme }) => theme.palette.primary.contrast};
  cursor: pointer;
  font-family: ${({ theme }) => theme.font.family};
  font-size: 16px;
  font-weight: ${({ isBold }) => (isBold ? "700" : "300")};
  line-height: 1.33;
  padding: ${({ isPadded }) => (isPadded ? "15px 42px" : "10px 20px")};
  transition: 0.4s cubic-bezier(0.25, 0.46, 0.45, 0.94);
  transition-property: background-color, color, opacity;

  &:hover {
    background: ${({ theme }) =>
      rgba(theme.palette.primary.main, BACKGROUND_ALPHA)};
  }
`
