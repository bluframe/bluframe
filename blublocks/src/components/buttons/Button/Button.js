// @flow
/* eslint-disable no-ternary */

import { Icon, IconEnd, IconStart, Wrapper } from "./styled"
import type { ComponentProps } from "."
import Loading from "components/Loading"
import React from "react"

const Button = ({
  bold,
  className,
  disabled,
  expanded,
  iconOnly,
  icon,
  iconEnd,
  iconStart,
  label,
  loading,
  noHoverHighlight,
  onClick,
  outlined,
  padded,
  rounded,
  secondary,
  small,
  transparent,
  type = "button"
}: ComponentProps): React$Node => (
  <Wrapper
    aria-disabled={disabled ? "true" : "false"}
    aria-label={iconOnly ? label : undefined}
    bold={bold}
    className={className}
    disabled={disabled}
    expanded={expanded}
    iconOnly={iconOnly}
    noHoverHighlight={noHoverHighlight}
    onClick={onClick}
    outlined={outlined}
    padded={padded}
    rounded={rounded}
    secondary={secondary}
    small={small}
    transparent={transparent}
    type={type}
  >
    {iconOnly ? <Icon>{icon}</Icon> : null}
    {!iconOnly && iconStart ? <IconStart>{iconStart}</IconStart> : null}
    {iconOnly ? null : label}
    {!iconOnly && iconEnd ? <IconEnd>{iconEnd}</IconEnd> : null}
    {loading ? (
      <IconEnd>
        <Loading contrast={!outlined} />
      </IconEnd>
    ) : null}
  </Wrapper>
)

export default Button
