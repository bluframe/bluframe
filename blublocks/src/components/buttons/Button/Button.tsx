/* eslint-disable no-ternary, max-lines-per-function, no-nested-ternary */

import { ComponentProps } from "."
import Loading from "components/Loading"
import { Wrapper } from "./styled"

const Button = ({
  ariaLabel,
  bold,
  className,
  disabled,
  expanded,
  iconOnly,
  icon,
  iconEnd,
  iconStart,
  label,
  loading,
  noHoverHighlight,
  onClick,
  outlined,
  padded,
  raised,
  rounded,
  secondary,
  small,
  transparent,
  type = "button"
}: ComponentProps) => (
  <Wrapper
    $bold={bold}
    $expanded={expanded}
    $hasIconEnd={Boolean(iconEnd)}
    $hasIconStart={Boolean(iconStart)}
    $iconOnly={iconOnly}
    $loading={loading}
    $noHoverHighlight={noHoverHighlight}
    $outlined={outlined}
    $padded={padded}
    $raised={raised}
    $rounded={rounded}
    $secondary={secondary}
    $small={small}
    $transparent={transparent}
    aria-disabled={disabled ? "true" : "false"}
    aria-label={ariaLabel ? ariaLabel : iconOnly ? label : undefined}
    className={className}
    disabled={disabled}
    onClick={onClick}
    type={type}
  >
    {iconOnly ? icon : null}
    {!iconOnly && iconStart ? iconStart : <span />}
    {iconOnly ? null : label}
    {!iconOnly && iconEnd ? iconEnd : <span />}
    {loading ? <Loading contrast={!outlined} /> : null}
  </Wrapper>
)

export default Button
