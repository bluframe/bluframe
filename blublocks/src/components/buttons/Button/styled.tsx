/* eslint-disable no-ternary, no-nested-ternary */

import Reset from "components/buttons/Reset"
import { Theme } from "styles/theme"
import colors from "styles/colors"
import getColorGrades from "helpers/getColorGrades"
import { rgba } from "polished"
import styled from "styled-components"

type Props = {
  $bold?: boolean
  $expanded?: boolean
  $hasIconEnd: boolean
  $hasIconStart: boolean
  $iconOnly?: boolean
  $loading?: boolean
  $noHoverHighlight?: boolean
  $outlined?: boolean
  $padded?: boolean
  $raised?: boolean
  $rounded?: boolean
  $secondary?: boolean
  $small?: boolean
  $transparent?: boolean
  disabled?: boolean
}

const BACKGROUND_ALPHA = 0.7
const BOX_SHADOW_ALPHA_ONE = 0.2
const BOX_SHADOW_ALPHA_TWO = 0.14
const BOX_SHADOW_ALPHA_THREE = 0.12

const getBackground = (props: Props & { theme: Theme }) => {
  if (props.$transparent) {
    return "transparent"
  }

  const colorGrades = getColorGrades(props)

  return props.$outlined ? colorGrades.contrast : colorGrades.main
}

const getBorder = (props: Props & { theme: Theme }) => {
  if (props.$transparent) {
    return "1px solid transparent"
  }

  const colorGrades = getColorGrades(props)

  return `1px solid ${colorGrades.main}`
}

const getColor = (props: Props & { theme: Theme }) => {
  const colorGrades = getColorGrades(props)

  return props.$outlined || props.$transparent
    ? colorGrades.main
    : colorGrades.contrast
}

export const Wrapper = styled(Reset).attrs((props: Props) => ({
  "data-variant": props.$secondary
    ? "secondary"
    : props.$outlined
    ? "outlined"
    : props.$iconOnly
    ? "iconOnly"
    : undefined
}))`
  align-items: center;
  background: ${getBackground};
  border: ${getBorder};
  border-radius: ${({ $rounded }: { $rounded?: boolean }) =>
    $rounded ? "50%" : "3px"};
  ${({ $raised }: { $raised?: boolean }) =>
    $raised &&
    `box-shadow: 0px 3px 1px -2px ${rgba(
      colors.black,
      BOX_SHADOW_ALPHA_ONE
    )}, 0px 2px 2px 0px ${rgba(
      colors.black,
      BOX_SHADOW_ALPHA_TWO
    )}, 0px 1px 5px 0px ${rgba(colors.black, BOX_SHADOW_ALPHA_THREE)};`}
  color: ${getColor};
  column-gap: 10px;
  cursor: ${({ disabled }: { disabled?: boolean }) =>
    disabled ? "not-allowed" : "pointer"};
  display: grid;
  font-family: ${({ theme }: { theme: Theme }) => theme.font.family};
  font-size: ${({ $small }: { $small?: boolean }) =>
    $small ? "14px" : "16px"};
  font-weight: ${({ $bold }: { $bold?: boolean }) => ($bold ? "700" : "300")};
grid-template-columns: ${({ $hasIconEnd, $hasIconStart, $loading }: Props) =>
  $loading
    ? "auto auto auto auto"
    : $hasIconEnd || $hasIconStart
    ? "auto auto auto"
    : "auto"};
  justify-items: center;
  line-height: 1.33;
      padding: ${({ $padded, $rounded, $small, $iconOnly }: Props) =>
        $small
          ? "6px 12px"
          : $iconOnly || $rounded
          ? "10px"
          : $padded
          ? "15px 42px"
          : "10px 20px"};
  transition: 0.4s cubic-bezier(0.25, 0.46, 0.45, 0.94);
  transition-property: background-color, color, opacity;
  width: ${({ $expanded }: { $expanded?: boolean }) =>
    $expanded ? "100%" : "auto"};

  &:hover {
  ${(props: any) =>
    !props.$noHoverHighlight &&
    `background:
      ${
        props.disabled || props.$transparent
          ? getBackground(props)
          : rgba(getBackground(props), BACKGROUND_ALPHA)
      };}`}

  & i,
  svg {
    color: ${getColor};
  }
`
