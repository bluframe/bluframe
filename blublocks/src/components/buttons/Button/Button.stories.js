// @flow

import Button, { type Props } from "components/buttons/Button"
import styled, { type StyledComponent } from "styled-components"
import Icon from "components/Icon"
import React from "react"
import type { Story } from "@storybook/react"
import type { Theme } from "@bluframe/blublocks"
import mediaQueries from "styles/mediaQueries"

const stories = {
  argTypes: {
    bold: { control: { type: "boolean" } },
    disabled: { control: { type: "boolean" } },
    expanded: { control: { type: "boolean" } },
    loading: { control: { type: "boolean" } },
    onClick: { action: "button click" },
    outlined: { control: { type: "boolean" } },
    padded: { control: { type: "boolean" } },
    secondary: { control: { type: "boolean" } },
    small: { control: { type: "boolean" } }
  },
  component: Button,
  title: "Buttons/Button"
}

const Template = (args: Props) => <Button {...args} />

const ExpandedWrapper: StyledComponent<
  {||},
  Theme,
  HTMLDivElement
> = styled.div`
  width: 300px;

  ${mediaQueries.tablet} {
    width: 300px;
  }
`

// eslint-disable-next-line react/no-multi-comp
const ExpandedTemplate = (args: Props) => (
  <ExpandedWrapper>
    <Button {...args} />
  </ExpandedWrapper>
)

export const Default: Story<Props> = Template.bind({})
Default.args = {
  label: "Button"
}

export const Bold: Story<Props> = Template.bind({})
Bold.args = {
  bold: true,
  label: "Button"
}

export const Disabled: Story<Props> = Template.bind({})
Disabled.args = {
  disabled: true,
  label: "Button"
}

export const Loading: Story<Props> = Template.bind({})
Loading.args = {
  label: "Button",
  loading: true
}

export const IconEnd: Story<Props> = Template.bind({})
IconEnd.args = {
  iconEnd: <Icon name="IoArrowForward" />,
  label: "Button"
}

export const IconOnly: Story<Props> = Template.bind({})
IconOnly.args = {
  icon: <Icon name="IoArrowBack" size="24px" />,
  iconOnly: true,
  label: "Go back"
}

export const IconStart: Story<Props> = Template.bind({})
IconStart.args = {
  iconStart: <Icon name="IoArrowBack" />,
  label: "Button"
}

export const NoHoverHighlight: Story<Props> = Template.bind({})
NoHoverHighlight.args = {
  label: "Button",
  noHoverHighlight: true
}

export const Outlined: Story<Props> = Template.bind({})
Outlined.args = {
  label: "Button",
  outlined: true
}

export const Padded: Story<Props> = Template.bind({})
Padded.args = {
  label: "Button",
  padded: true
}

export const Expanded: Story<Props> = ExpandedTemplate.bind({})
Expanded.args = {
  expanded: true,
  label: "Button"
}

export const Secondary: Story<Props> = Template.bind({})
Secondary.args = {
  label: "Button",
  secondary: true
}

export const Small: Story<Props> = Template.bind({})
Small.args = {
  label: "Button",
  small: true
}

export const TransparentIconOnly: Story<Props> = Template.bind({})
TransparentIconOnly.args = {
  icon: <Icon name="IoArrowBack" size="24px" />,
  iconOnly: true,
  label: "Go back",
  transparent: true
}

export default stories
