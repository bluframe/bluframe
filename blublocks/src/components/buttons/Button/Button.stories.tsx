import Button, { type Props } from "components/buttons/Button"
import Icon from "components/Icon"
import { StoryObj } from "@storybook/react"
import mediaQueries from "styles/mediaQueries"
import styled from "styled-components"

const meta = {
  argTypes: {
    bold: { control: { type: "boolean" } },
    disabled: { control: { type: "boolean" } },
    expanded: { control: { type: "boolean" } },
    loading: { control: { type: "boolean" } },
    onClick: { action: "button click" },
    outlined: { control: { type: "boolean" } },
    padded: { control: { type: "boolean" } },
    secondary: { control: { type: "boolean" } },
    small: { control: { type: "boolean" } }
  },
  component: Button,
  title: "Buttons/Button"
}

type Story = StoryObj<typeof Button>

const ExpandedWrapper = styled.div`
  width: 300px;

  ${mediaQueries.tablet} {
    width: 300px;
  }
`

// eslint-disable-next-line react/no-multi-comp
const ExpandedTemplate = (args: Props) => (
  <ExpandedWrapper>
    <Button {...args} />
  </ExpandedWrapper>
)

export const Default: Story = {
  args: {
    label: "Button"
  }
}

export const Bold: Story = {
  args: {
    bold: true,
    label: "Button"
  }
}

export const Disabled: Story = {
  args: {
    disabled: true,
    label: "Button"
  }
}

export const Loading: Story = {
  args: {
    label: "Button",
    loading: true
  }
}

export const IconEnd: Story = {
  args: {
    iconEnd: <Icon name="IoArrowForward" />,
    label: "Button"
  }
}

export const IconOnly: Story = {
  args: {
    icon: <Icon name="IoArrowBack" size="24px" />,
    iconOnly: true,
    label: "Go back"
  }
}

export const IconOnlyRaisedAndRounded: Story = {
  args: {
    icon: <Icon name="IoArrowBack" size="24px" />,
    iconOnly: true,
    label: "Go back",
    raised: true,
    rounded: true
  }
}

export const IconOnlyRounded: Story = {
  args: {
    icon: <Icon name="IoArrowBack" size="24px" />,
    iconOnly: true,
    label: "Go back",
    rounded: true
  }
}

export const IconStart: Story = {
  args: {
    iconStart: <Icon name="IoArrowBack" />,
    label: "Button"
  }
}

export const NoHoverHighlight: Story = {
  args: {
    label: "Button",
    noHoverHighlight: true
  }
}

export const Outlined: Story = {
  args: {
    label: "Button",
    outlined: true
  }
}

export const Padded: Story = {
  args: {
    label: "Button",
    padded: true
  }
}

export const Expanded: Story = {
  args: {
    expanded: true,
    label: "Button"
  },
  render: ExpandedTemplate
}

export const Secondary: Story = {
  args: {
    label: "Button",
    secondary: true
  }
}

export const Small: Story = {
  args: {
    label: "Button",
    small: true
  }
}

export const TransparentIconOnly: Story = {
  args: {
    icon: <Icon name="IoArrowBack" size="24px" />,
    iconOnly: true,
    label: "Go back",
    transparent: true
  }
}

export default meta
