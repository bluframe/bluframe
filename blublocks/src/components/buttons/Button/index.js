// @flow

import Button from "./Button"
import type { ButtonProps } from "@bluframe/blublocks"

export type ComponentProps = {|
  ...ButtonProps
|}

export default Button
