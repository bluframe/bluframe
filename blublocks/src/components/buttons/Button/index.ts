import React, { MouseEventHandler } from "react"
import Button from "./Button"

export interface Props {
  ariaLabel?: string
  bold?: boolean
  className?: string
  disabled?: boolean
  expanded?: boolean
  icon?: React.ReactNode
  iconEnd?: React.ReactNode
  iconOnly?: boolean
  iconStart?: React.ReactNode
  label: string
  loading?: boolean
  noHoverHighlight?: boolean
  onClick?: MouseEventHandler<HTMLButtonElement>
  outlined?: boolean
  padded?: boolean
  raised?: boolean
  rounded?: boolean
  secondary?: boolean
  small?: boolean
  transparent?: boolean
  type?: "button" | "reset" | "submit"
}

export interface ComponentProps extends Props {}

export default Button
