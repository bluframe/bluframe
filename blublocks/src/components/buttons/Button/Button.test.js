// @flow
/* eslint-disable max-lines-per-function, max-statements */

import { fireEvent, render } from "tests"
import Button from "./Button"
import { type ComponentProps } from "."
import React from "react"

jest.mock(".", () => ({}))
jest.mock("react-markdown", () => "ReactMarkdown")

describe("Button", () => {
  const props: ComponentProps = {
    label: "Button",
    onClick: jest.fn()
  }

  it("renders", () => {
    const { container } = render(<Button {...props} />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders bold", () => {
    const { container } = render(<Button {...props} bold />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders padded", () => {
    const { container } = render(<Button {...props} padded />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders small", () => {
    const { container } = render(<Button {...props} small />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders disabled", () => {
    const { container } = render(<Button {...props} disabled />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders outlined", () => {
    const { container } = render(<Button {...props} outlined />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders secondary", () => {
    const { container } = render(<Button {...props} secondary />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it("renders with iconStart and iconEnd", () => {
    const { container } = render(
      <Button
        {...props}
        iconEnd={<div>{"Icon End"}</div>}
        iconStart={<div>{"Icon Start"}</div>}
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it("triggers onClick event", () => {
    const { getByText } = render(<Button {...props} />)

    fireEvent.click(getByText(props.label))
    expect(props.onClick).toHaveBeenCalled()
  })

  it("does not trigger onClick event if disabled", () => {
    const { getByText } = render(<Button {...props} disabled />)

    fireEvent.click(getByText(props.label))
    expect(props.onClick).not.toHaveBeenCalled()
  })
})
