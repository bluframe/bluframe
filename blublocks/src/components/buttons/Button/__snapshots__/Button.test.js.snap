// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders 1`] = `
<button
  aria-disabled="false"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb fnYKKn"
  type="button"
>
  Button
</button>
`;

exports[`Button renders bold 1`] = `
<button
  aria-disabled="false"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb hIHjsH"
  type="button"
>
  Button
</button>
`;

exports[`Button renders disabled 1`] = `
<button
  aria-disabled="true"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb ibyxDU"
  disabled=""
  type="button"
>
  Button
</button>
`;

exports[`Button renders iconOnly button 1`] = `
<button
  aria-disabled="false"
  aria-label="Button"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb fnYKKn"
  type="button"
>
  <span
    class="styled__Icon-sc-pzlzbh-1"
  >
    <span
      class="styled__Icon-sc-ztr6p6-0 kwuIiV"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M244 400L100 256l144-144M120 256h292"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="48"
        />
      </svg>
    </span>
  </span>
</button>
`;

exports[`Button renders outlined 1`] = `
<button
  aria-disabled="false"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb dNGKMI"
  type="button"
>
  Button
</button>
`;

exports[`Button renders padded 1`] = `
<button
  aria-disabled="false"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb kIBhZW"
  type="button"
>
  Button
</button>
`;

exports[`Button renders secondary 1`] = `
<button
  aria-disabled="false"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb ztbDP"
  type="button"
>
  Button
</button>
`;

exports[`Button renders small 1`] = `
<button
  aria-disabled="false"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb bxTKRn"
  type="button"
>
  Button
</button>
`;

exports[`Button renders transparent iconOnly button 1`] = `
<button
  aria-disabled="false"
  aria-label="Button"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb jkrxpI"
  type="button"
>
  <span
    class="styled__Icon-sc-pzlzbh-1"
  >
    <span
      class="styled__Icon-sc-ztr6p6-0 kwuIiV"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M244 400L100 256l144-144M120 256h292"
          fill="none"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="48"
        />
      </svg>
    </span>
  </span>
</button>
`;

exports[`Button renders with iconStart and iconEnd 1`] = `
<button
  aria-disabled="false"
  class="Reset__Button-sc-ohlhxz-0 styled__Wrapper-sc-pzlzbh-0 eNpAQb fnYKKn"
  type="button"
>
  <span
    class="styled__IconStart-sc-pzlzbh-3"
  >
    <div>
      Icon Start
    </div>
  </span>
  Button
  <span
    class="styled__IconEnd-sc-pzlzbh-2"
  >
    <div>
      Icon End
    </div>
  </span>
</button>
`;
